### PR TITLE
Stop running iterators when js client disconnects

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

If you run an iterator event and close the page the python function will keep running in the backend.

To test, run the following demo, click the button and close the page. Things will still be being printed to the console in main. On this branch, they do not. Also added a playwright test for this.

```python
    def iterate():
        for i in range(8):
            print(f"UPDATE: {i}")
            yield i
            time.sleep(0.5)

    with gr.Blocks() as demo:
        with gr.Row():
            c_time2 = gr.Textbox(label="Iteration Step")
            btn = gr.Button("Start")
            btn.click(iterate, None, c_time2, api_name="start")


   demo.launch()
```

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: demo/cancel_events/run.py
Comment (abidlabs) [inline demo/cancel_events/run.py:1]: Let's revert these changes?

Comment (abidlabs) [reply demo/cancel_events/run.py:1]: Ah got it! We don't use this demo in any of our docs so it should be fine

File: demo/cancel_events/run.py [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
+import pathlib
 
+log_file = (pathlib.Path(__file__).parent / "cancel_events_output_log.txt").resolve()
 
 def fake_diffusion(steps):
+    log_file.write_text("")
     for i in range(steps):
         print(f"Current step: {i}")
+        log_file.write_text(f"Current step: {i}\n")
         time.sleep(0.5)
```
Comment (abidlabs) [inline demo/cancel_events/run.py:13]: Could we make this faster? So that we could reduce the time that we have to wait in the test
```suggestion
        time.sleep(0.2)
```

File: gradio/routes.py [hunk 2]
```diff
@@ -803,11 +803,11 @@ async def sse_stream(request: fastapi.Request):
                         message=str(e),
                     )
                     response = process_msg(message)
-                    if response is not None:
```
Comment (abidlabs) [inline gradio/routes.py:806]: Ah so we already had the deletion logic in place, but it was never called?

File: js/app/test/cancel_events.spec.ts [hunk 3]
```diff
@@ -44,3 +45,26 @@ test("when using an iterative function it should be possible to cancel the funct
...
+	// wait for the duration of the entire iteration
+	// check that the final value did not get written
+	// to the log file. That's our proof python stopped
+	// running
+	await new Promise((resolve) => setTimeout(resolve, 5000));
+	const data = readFileSync(
+		"../../demo/cancel_events/cancel_events_output_log.txt",
+		"utf-8"
+	);
+	expect(data).not.toContain("Current step: 8");
```
Comment (abidlabs) [inline js/app/test/cancel_events.spec.ts:69]: Let's also ensure that at least one of the steps that should be logged are logged
```suggestion
	expect(data).not.toContain("Current step: 1");
	expect(data).not.toContain("Current step: 8");
```

File: gradio/routes.py (hunk 2 above)
Comment (freddyaboulton) [reply gradio/routes.py:806]: Yea we yielded the response before the clean up logic happened so we never reached that point in the code

Comment (abidlabs) [reply gradio/routes.py:806]: Nice catch

**Review by abidlabs — APPROVED** [review]
LGTM with a few suggestions. Cool test!

File: js/app/test/cancel_events.spec.ts (hunk 3 above)
Comment (freddyaboulton) [reply js/app/test/cancel_events.spec.ts:70]: Reasonable!

**Comment by freddyaboulton** [comment]
Thanks for the review @abidlabs ! I addressed your comments so will merge once ci passes
